pkg/collector/container.go

```diff
@@ -15,24 +15,56 @@
 package collector
 
 import (
+	"errors"
 	"fmt"
+	"net"
+	"sort"
+	"strconv"
 
 	"github.com/go-logr/logr"
+	"github.com/mitchellh/mapstructure"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
 
 	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
 	"github.com/open-telemetry/opentelemetry-operator/internal/config"
 	"github.com/open-telemetry/opentelemetry-operator/pkg/collector/adapters"
 	"github.com/open-telemetry/opentelemetry-operator/pkg/naming"
 )
 
+const maxPortLen = 15
+
+var errInvalidPort = errors.New("invalid port name/num")
+
 // Container builds a container for the given collector.
 func Container(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelemetryCollector) corev1.Container {
 	image := otelcol.Spec.Image
 	if len(image) == 0 {
 		image = cfg.CollectorImage()
 	}
 
+	// build container ports from service ports
+	ports := getConfigContainerPorts(logger, otelcol.Spec.Config)
+	for _, p := range otelcol.Spec.Ports {
```

Comment:
This function does a couple of different things:
1. name truncation
2. validation
3. create a container port

The truncation and validation should be done in the defaulting and validating webhook.  The otelcol spec received in this function should be valid and prepared for use.

Reply:
The validating webhook can as well return user-facing error if the name validation fails.

Reply:
I can't add the current truncation to the defaulting webhook because it causes an import cycle:
```
package github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1
	imports github.com/open-telemetry/opentelemetry-operator/pkg/naming
	imports github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1: import cycle not allowed
```
Is it reasonable for us to error out when a Spec Port is too long (like any other port naming error)? Or should I truncate a different way?